src/actions/receive-profile.js

```diff
@@ -431,52 +436,160 @@ async function _fetchProfile({ url, onTemporaryError }: FetchProfileArgs) {
+    );
+  }
+}
 
-      dispatch(receiveProfileFromStore(profile));
-    } catch (error) {
-      dispatch(fatalErrorReceivingProfileFromStore(error));
+/**
+ * Don't trust third party responses, try and handle a variety of responses gracefully.
+ */
+async function _extractJsonFromResponse(
+  response: Response,
+  reportError: Function,
+  fileType: 'application/json' | null
+): Promise<any> {
+  try {
+    // Don't check the content-type, but attempt to parse the response as JSON.
+    return await response.json();
+  } catch (error) {
+    // Change the error message depending on the circumstance:
+    let message;
+    if (fileType === 'application/json') {
+      message = 'The profile’s JSON could not be decoded.';
+    } else {
+      message = oneLine`
+        The profile could not be decoded. This does not look like a supported file
+        type.
+      `;
     }
-  };
+
+    // Provide helpful debugging information to the console.
+    reportError(message);
+    reportError('JSON parsing error:', error);
+    reportError('Fetch response:', response);
+
+    throw new Error(
+      `${message} The full error information has been printed out to the DevTool’s console.`
+    );
+  }
 }
 
-export function retrieveProfileFromUrl(
+export function retrieveProfileFromStore(
+  hash: string
+): ThunkAction<Promise<void>> {
+  return retrieveProfileOrZipFromUrl(
+    `https://profile-store.commondatastorage.googleapis.com/${hash}`
+  );
+}
+
+/**
+ * Runs a fetch on a URL, and downloads the file. If it's JSON, then it attempts
+ * to process the profile. If it's a zip file, it tries to unzip it, and save it
+ * into the store so that the user can then choose which file to load.
+ */
+export function retrieveProfileOrZipFromUrl(
   profileUrl: string
 ): ThunkAction<Promise<void>> {
   return async function(dispatch) {
     dispatch(waitingForProfileFromUrl());
 
     try {
-      const serializedProfile = await _fetchProfile({
+      const response = await _fetchProfile({
         url: profileUrl,
         onTemporaryError: (e: TemporaryError) => {
           dispatch(temporaryErrorReceivingProfileFromUrl(e));
         },
       });
 
-      const profile = unserializeProfileOfArbitraryFormat(serializedProfile);
-      if (profile === undefined) {
-        throw new Error('Unable to parse the profile.');
-      }
+      const serializedProfile = response.profile;
+      const zip = response.zip;
+      if (serializedProfile) {
+        const profile = unserializeProfileOfArbitraryFormat(serializedProfile);
+        if (profile === undefined) {
+          throw new Error('Unable to parse the profile.');
+        }
 
-      dispatch(receiveProfileFromUrl(profile));
+        dispatch(viewProfile(profile));
+      } else if (zip) {
+        dispatch(receiveZipFile(zip));
+      } else {
+        throw new Error(
+          'Expected to receive a zip file or profile from _fetchProfile.'
```

Comment:
I'm not finding this level of fine-grained grammatical feedback helpful. I think it's detracting from the speed at which we can move forward with projects. To me, it still reads as grammatically correct.f

Let me know if you want me to still change it and I can.

Reply:
Well, it's not a big deal, in one way or another. It's more that I skimmed through this code and thought that while we're here this can be changed. I don't mind if you don't want to change it (I mean it, I _really_ don't mind !). 

But I think you spent more time writing this comment, and maybe dealing with some unwanted frustration, than if you just changed it without an afterthought...